Dockerfile

```diff
@@ -13,46 +13,179 @@
+ENV DEBIAN_FRONTEND=noninteractive LANGUAGE=C.UTF-8 LANG=C.UTF-8 LC_ALL=C.UTF-8 \
+    LC_CTYPE=C.UTF-8 LC_MESSAGES=C.UTF-8
+
+# Increase the value below to force renstalling of all dependencies
+ENV DEPENDENCIES_EPOCH_NUMBER=${DEPENDENCIES_EPOCH_NUMBER}
+
+# Install curl and gnupg2 - needed to download nodejs in next step
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+           curl gnupg2 \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+
+# Install basic apt dependencies
+RUN curl -sL https://deb.nodesource.com/setup_10.x | bash - \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends \
+           # Packages to install \
+           libsasl2-dev freetds-bin build-essential sasl2-bin \
+           libsasl2-2 libsasl2-dev libsasl2-modules \
+           default-libmysqlclient-dev apt-utils curl rsync netcat locales  \
+           freetds-dev libkrb5-dev libssl-dev libffi-dev libpq-dev git \
+           nodejs gosu sudo \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN adduser airflow \
+    && echo "airflow ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers.d/airflow \
+    && chmod 0440 /etc/sudoers.d/airflow
+
+############################################################################################################
+# This is the target image - it installs PIP and NPN dependencies including efficient caching
+# mechanisms - it might be used to build the bare airflow build or CI build
+# Parameters:
+#    APT_DEPS_IMAGE - image with APT dependencies. It might either be base deps image with airflow
+#                     dependencies or CI deps image that contains also CI-required dependencies
+############################################################################################################
+FROM airflow-apt-deps as main
+
+SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]
+
+WORKDIR /opt/airflow
+
+RUN echo "Airflow version: ${AIRFLOW_VERSION}"
+
+ARG AIRFLOW_HOME=/opt/airflow
+ENV AIRFLOW_HOME=${AIRFLOW_HOME}
 
-FROM python:3.6-slim
-SHELL ["/bin/bash", "-xc"]
+RUN mkdir -pv ${AIRFLOW_HOME} \
+    && chown -R airflow.airflow ${AIRFLOW_HOME}
 
-ENV AIRFLOW_HOME=/usr/local/airflow
-ARG AIRFLOW_DEPS="all"
-ARG PYTHON_DEPS=""
-ARG BUILD_DEPS="freetds-dev libkrb5-dev libssl-dev libffi-dev libpq-dev git"
-ARG APT_DEPS="libsasl2-dev freetds-bin build-essential default-libmysqlclient-dev apt-utils curl rsync netcat locales"
+# Increase the value here to force reinstalling Apache Airflow pip dependencies
+ENV FORCE_REINSTALL_ALL_PIP_DEPENDENCIES=1
 
-ENV PATH="$HOME/.npm-packages/bin:$PATH"
+# Optimizing installation of Cassandra driver
+ARG CASS_DRIVER_BUILD_CONCURRENCY
+ARG CASS_DRIVER_NO_CYTHON
+ENV CASS_DRIVER_BUILD_CONCURRENCY=${CASS_DRIVER_BUILD_CONCURRENCY}
+ENV CASS_DRIVER_NO_CYTHON=${CASS_DRIVER_NO_CYTHON}
```

Comment:
Do we  need  these here given they are  in the base layer?

Reply:
Very good question. Actually yes and no :).

That's one of the things I've learned about some intricacies of multi-stage builds. You need to declare ARGs in the other image if you want to use it. Otherwise the ARGS will be empty. Also I found that the default value from the base image carries over to the second definition (surprisingly) so you do not have repeat the default value :D. 

Then I decided then that all ARGS with default parameters will be defined at the beginning of the base layer (it's easier to find them there) and used when they are needed in the multi-stage images. Some duplication on one hand but then all the "configuration" of the image is defined in one place.

Alternative (which  have to test) is to move the CASS_ ARGs (and few others such as no PIP_CACHE) with defaults to be only defined in the "main" layer and not at the top of the file. I think that could work and is a bit less duplication more distribution of configuration. I am on the fence on this one.

What do you think? What's your subjective opinion on that?

Reply:
Actually I gave it a spin and tried to move the default definition closer to the actual use and it feels a bit better. Having defaults close to where they are used is actually much more readable and there is little need to group ARGS together at the beginning of the Dockerfile.

The only one that needs to be repeated is PYTHON_BASE_IMAGE - it has to be set before FROM and needs to be repeated after FROM because otherwise the ARG value is not visible for the ENV line (strange and not intuitive but this is how it works for Dockerfile - just double-checked)

<img width="655" alt="Screenshot 2019-03-25 at 23 10 35" src="https://user-images.githubusercontent.com/595491/54957561-5e086880-4f53-11e9-936a-b7a02b597946.png">

<img width="663" alt="Screenshot 2019-03-25 at 23 11 27" src="https://user-images.githubusercontent.com/595491/54957566-61035900-4f53-11e9-935c-565dd6258d08.png">

Reply:
Fixed. We will have far less duplication now in the next pushed version